x-pack/plugins/siem/public/components/super_date_picker/index.tsx

```diff
@@ -0,0 +1,227 @@
+  start: string;
+  end: string;
+}
+
+export type SuperDatePickerProps = OwnProps &
+  SuperDatePickerDispatchProps &
+  SuperDatePickerStateRedux;
+
+export interface SuperDatePickerState {
+  isAutoRefreshOnly: boolean;
+  isPaused: boolean;
+  isQuickSelection: boolean;
+  recentlyUsedRanges: TimeArgs[];
+  refreshInterval: number;
+  showUpdateButton: boolean;
+}
+
+export const SuperDatePickerComponent = class extends Component<
+  SuperDatePickerProps,
+  SuperDatePickerState
+> {
+  constructor(props: SuperDatePickerProps) {
+    super(props);
+
+    this.state = {
+      isAutoRefreshOnly: false,
+      isPaused: true,
+      isQuickSelection: true,
+      recentlyUsedRanges: [],
+      refreshInterval: 300000,
+      showUpdateButton: true,
+    };
+  }
+
+  public render() {
+    const { end, start, kind, option, isLoading } = this.props;
+    let endDate = new Date(end).toISOString();
+    let startDate = new Date(start).toISOString();
+    if (kind === 'relative') {
+      startDate = option;
+      endDate = 'now';
+    }
+    return (
+      <MyEuiSuperDatePicker
+        end={endDate}
+        isLoading={isLoading}
+        isPaused={this.state.isPaused}
+        onTimeChange={this.onTimeChange}
+        onRefreshChange={this.onRefreshChange}
+        onRefresh={this.onRefresh}
+        recentlyUsedRanges={this.state.recentlyUsedRanges}
+        refreshInterval={this.state.refreshInterval}
+        showUpdateButton={this.state.showUpdateButton}
+        start={startDate}
+      />
+    );
+  }
+  private onRefresh = ({ start, end, refreshInterval }: OnRefreshProps): void => {
+    this.updateReduxTime({
+      start,
+      end,
+      isQuickSelection: this.state.isQuickSelection,
+      isInvalid: false,
+    });
+    this.refetchQuery(this.props.refetch);
+  };
+
+  private onRefreshChange = ({ isPaused, refreshInterval }: OnRefreshChangeProps): void => {
+    const { id, duration, stopAutoReload, startAutoReload } = this.props;
+
+    if (duration !== refreshInterval) {
+      this.props.setDuration({ id, duration: refreshInterval });
+    }
+
+    if (isPaused) {
+      stopAutoReload({ id });
+    } else {
+      startAutoReload({ id });
+    }
+
+    if (!this.state.isQuickSelection) {
+      this.refetchQuery(this.props.refetch);
+    }
+
+    this.setState({
+      ...this.state,
+      isPaused,
+    });
+  };
+
+  private refetchQuery = (query: inputsModel.Refetch[]) => {
+    query.forEach((refetch: inputsModel.Refetch) => refetch());
+  };
+
+  private formatDate = (date: string) => {
+    const momentDate = dateMath.parse(date);
+    return momentDate ? momentDate.valueOf() : 0;
+  };
+
+  private onTimeChange = ({ start, end, isQuickSelection, isInvalid }: OnTimeChangeProps) => {
+    if (!isInvalid) {
+      this.updateReduxTime({ start, end, isQuickSelection, isInvalid });
+      this.setState((prevState: SuperDatePickerState) => {
+        let recentlyUsedRanges = prevState.recentlyUsedRanges.filter(
```

Comment:
consider eliminating both the re-assignment (`let`) and the conditional logic by using `lodash` [take](https://lodash.com/docs/#take):

```
        const recentlyUsedRanges = [
          { start, end },
          ...take(
            MAX_RECENTLY_USED_RANGES,
            prevState.recentlyUsedRanges.filter(
              recentlyUsedRange =>
                !(recentlyUsedRange.start === start && recentlyUsedRange.end === end)
            )
          ),
        ];
```